### PR TITLE
Add support for timers/promises module from nodejs

### PR DESCRIPTION
**Pull request by WhiteAutumn**
#### Purpose (TL;DR)

Fix issue #469 by implementing promise-based wrappers around setTimeout, setImmediate, and setInterval.

#### Solution
This PR takes a similar approach to #467 by @swenzel-arc of detecting the availability of the `timers/promises` module by attempting to import it and then overwriting the functions on the import object, which should be the same object users have, given they do not mess with the module cache. Like the aforementioned PR, the functions are only overwritten when calling `install()` on the `globalObject`.

#### Limitations
* When aborting a timer using an `AbortSignal`, the error thrown is not the same as the one thrown by node. This is due to the `AbortError` error being publicly available. See this [issue](https://github.com/nodejs/node/issues/38361).
* This PR does not touch [scheduler.wait](https://nodejs.org/api/timers.html#timerspromisesschedulerwaitdelay-options) or [scheduler.yield](https://nodejs.org/api/timers.html#timerspromisesscheduleryield) of the `timers/promises` module as they are marked as experimental.

### CODE REVIEW

File: src/fake-timers-src.js
Comment (fatso83) [inline src/fake-timers-src.js:1925]: Could use optional chaining on most of these options.signal blocks to collapse the indentation.
```suggestion
                                  options.signal.?removeEventListener.?(
                                      "abort",
                                      abort,
                                  );
```

**Review by fatso83 — COMMENTED** [review]
Very solid work! Love the implementation and the tests. Just had a minor comment. You can see what you mean before I approve. It's just a minor quibble.

File: src/fake-timers-src.js
Comment (WhiteAutumn) [reply src/fake-timers-src.js:1925]: Sure! I'm all for optional chaining. I'll go through and replace instances of `if (options.signal)` with optional chaining.

**Review by benjamingr — APPROVED** [review]
There are some leaks in the signal (e.g. we don't remove the event listener when fake timers are uninstalled) but generally that's fine for the first version so lgtm.

**Comment by fatso83** [comment]
Good note about the removal of event listeners. When I do any development on Sinon projects, I tend to make use of the watch feature of the test runner. That's where stuff like not cleaning up properly tends to show itself.

If it's not too much problem, I'd like to see that being done.

**Comment by WhiteAutumn** [comment]
Sure, I can take a look at it. Just so we're on the same page, we're talking about the scenario where while a timer is running, an abort signal has been passed, and the clock is uninstalled before the timer fires, leading to it never firing and the therefore the abort signal listener never being removed?

This is something I have missed, and the listeners should absolutely be removed in that scenario, but for what it's worth I don't think it would lead to memory leaks while running tests in watch mode. As long as no references remain to the abort signal, it should be garbage collected and the event listener along with it.

As to how to solve this, I would like a second opinion on how to approach it. The way I see it this would require something akin to an `onUninstall` event that the promise based implementations can listen to and then remove the abort event listeners when fired. It doesn't feel great adding more event listeners to solve hanging event listeners but I can't think of any other way of doing it. What do you think?

**Comment by fatso83** [comment]
You know what, I don't have available mental capacity (or time) at the moment to dig deeply into this, and if it really isn't an issue in practice, then I have no issues merging this now. There's always room for another PR later, if you find a nice way that isn't too complicated.

**Review by fatso83 — APPROVED** [review]

**Comment by fatso83** [comment]
Thanks for the reminder :)

**Comment by benjamingr** [comment]
> This is something I have missed, and the listeners should absolutely be removed in that scenario, but for what it's worth I don't think it would lead to memory leaks while running tests in watch mode. As long as no references remain to the abort signal, it should be garbage collected and the event listener along with it.

References can remain to the signal (e.g. the signal is the result of `AbortSignal.any` of a local signal and a process-level signal.

It's fine for it to happen at a follow up PR. Generally you did cleanup a lot better than I expected in this PR in the first place and it's a relatively small miss. I review a lot of AbortSignal library code so this is a compliment :)

> As to how to solve this, I would like a second opinion on how to approach it. The way I see it this would require something akin to an onUninstall event that the promise based implementations can listen to and then remove the abort event listeners when fired. It doesn't feel great adding more event listeners to solve hanging event listeners but I can't think of any other way of doing it. What do you think?

We keep a list/map or all signal and event listeners associated with the clock instance and when the clock is uninstalled we call that?

```js
// Somewhere when an event listener is added
signal.addEventListener('abort', listener, options); 
clock.listners.push({ signal, listener });

// In uninstall
clock.listeners.forEach(({ signal, listener }) => signal.removeEventListener(listener))
```

**Comment by WhiteAutumn** [comment]
> References can remain to the signal (e.g. the signal is the result of AbortSignal.any of a local signal and a process-level signal.
Ohh, I see. Yeah then it's definitely something that would be good to get fixed before a release.

> We keep a list/map or all signal and event listeners associated with the clock instance and when the clock is uninstalled we call that?
Thanks for the input. I'll proceed with that kind of a solution and make a separate PR.

**Comment by fatso83** [comment]
@u873838 Thanks to White Autumns swift second PR this is already out.

**Pull request merged**
